SceneInspector : add Subdivision entries

mattigruener opened the pull request:
Hey John,

Before heading out now, I spent a little time tonight looking at this one to get away from the non-coding stuff I was doing and to get the ball rolling on my last ticket in the creases mission. Could you take a look? Gives us something to talk about tomorrow if you get a chance to look at the code. If you don't that's fine :)

It's a little WIP because the dragging doesn't work yet. I believe the `PrimitiveVariables` have a similar problem? Mapping between cpp and python isn't working for `PrimitiveVariables` and my new type (which I've just jammed into the global namespace for now). Let's discuss how to fix that? I used a `namedtuple` because I was hoping the conversion would just work for a native type like that, but it didn't. I guess this is what you had in mind in general, though?

Thanks :)

Matti.


PS: I've been using this ugly bit of code for testing. Creates two planes with a crease each but differing sharpness values.

```
import IECoreScene
import GafferSceneTest

plane = IECoreScene.MeshPrimitive.createPlane( imath.Box2f( imath.V2f( -1 ), imath.V2f( 1 ) ) )
plane.setCreases( IECore.IntVectorData( [ 2 ] ), IECore.IntVectorData( [ 0, 1 ] ), IECore.FloatVectorData( [ 2.5 ] ) )
input2 = GafferSceneTest.CompoundObjectSource()
input2["in"].setValue(
	IECore.CompoundObject( {
		"bound" : IECore.Box3fData( plane.bound() ),
		"children" : {
			"planeGroup" : {
				"bound" : IECore.Box3fData( plane.bound() ),
				"children" : {
					"plane" : {
						"bound" : IECore.Box3fData( plane.bound() ),
						"object" : plane,
					},
				},
			},
		},
	} ),
)

root.addChild( input2 )

plane = IECoreScene.MeshPrimitive.createPlane( imath.Box2f( imath.V2f( -1 ), imath.V2f( 1 ) ) )
plane.setCreases( IECore.IntVectorData( [ 2 ] ), IECore.IntVectorData( [ 0, 1 ] ), IECore.FloatVectorData( [ 1 ] ) )
input2 = GafferSceneTest.CompoundObjectSource()
input2["in"].setValue(
	IECore.CompoundObject( {
		"bound" : IECore.Box3fData( plane.bound() ),
		"children" : {
			"planeGroup" : {
				"bound" : IECore.Box3fData( plane.bound() ),
				"children" : {
					"plane" : {
						"bound" : IECore.Box3fData( plane.bound() ),
						"object" : plane,
					},
				},
			},
		},
	} ),
)

root.addChild( input2 )

```

johnhaddon commented:
Thanks Matti! This looks like what I had in mind, and you're right about the dragging problem. The issue is that [the data](https://github.com/GafferHQ/gaffer/blob/master/include/GafferUI/DragDropEvent.h#L65) for DragDropEvent has to be `RunTimeTyped`, so C++ and Python code can interoperate with the same events. I think perhaps the best way around that for now is for the inspector to return a CompoundData encoding the creases/corners. This will also have the benefit that we can simply omit the lengths when returning corners.

The downside is that we won't have a unique type to detect in `__formatValues()`, but I don't think that's too bad. At some point we will need to extract the core UI components from the SceneInspector so that they can be reused in an ImageInspector, and `__formatValues()` will need to be cleansed of its scene-specific formatters. Let's discuss approaches to that in our chat later today...

mattigruener commented:
Updated. What do you think? :) I think what I have just added is pretty close to what we discussed today.

- Is there any situation in which the less explicit typing could cause an issue?
- How concerned are we about potentially dragging a gazillion "P" values around now that the PrimitiveVariables are draggable?

johnhaddon commented on python/GafferSceneUI/SceneInspector.py:
```diff
@@ -482,7 +482,7 @@ def _formatValues( self, values ) :
 		elif isinstance( values[0], basestring ) :
 			return self.__formatStrings( [ str( v ) for v in values ] )
-		elif isinstance( values[0], IECoreScene.PrimitiveVariable ) :
+		elif isinstance( values[0], IECore.CompoundData ) :
```
Let's do as we did for the creases, and use a custom TextDiff subclass instead of have TextDiff know about primitive variables. That way we're one step closer to being able to move TextDiff to GafferUI.

johnhaddon commented on python/GafferSceneUI/SceneInspector.py:
```diff
@@ -680,6 +680,27 @@ def __dragEnd( self, widget, event ) :
 
 	__htmlFooter = "</body></html>"
 
+class SubdivisionTextDiff( TextDiff ) :
```
Oh, and give it an `_` prefix to make it clear its for internal use.

mattigruener commented on python/GafferSceneUI/SceneInspector.py:
```diff
@@ -680,6 +680,27 @@ def __dragEnd( self, widget, event ) :
+
+		if len( values ) == 0 :
+			return []
+		elif isinstance( values[0], IECore.CompoundData ) :
```
I think I'm in favor of moving it. I've just updated the PR with a commit that addresses the rest of the feedback and if we want to move `interpolation` I could add that and squash in the commit if you're ok with what I've done there.

Thanks :)

johnhaddon replied:
Looks good to me - if you could shift the interpolation over and squash that would be great...

mattigruener commented:
Updated.

I've capitalized "interpolation" but otherwise this should just address your feedback and move that property over to the new Subdivision section.

To test the CurvePrimitive bits I've used this code:

```
import IECoreScene
import GafferSceneTest

testObject = IECoreScene.CurvesPrimitive(

	IECore.IntVectorData( [ 7, 7 ] ),
	IECore.CubicBasisf.bezier(),
	False,
	IECore.V3fVectorData(
		[
			imath.V3f( 0, 0, 0 ),
			imath.V3f( 0, 1, 0 ),
			imath.V3f( 1, 1, 0 ),
			imath.V3f( 1, 0, 0 ),
			imath.V3f( 1, -1, 0 ),
			imath.V3f( 2, -1, 0 ),
			imath.V3f( 2, 0, 0 ),

			imath.V3f( 0, 0, 0 ),
			imath.V3f( 0, 0, 1 ),
			imath.V3f( 1, 0, 1 ),
			imath.V3f( 1, 0, 0 ),
			imath.V3f( 1, 0, -1 ),
			imath.V3f( 2, 0, -1 ),
			imath.V3f( 2, 0, 0 )
		]
	)
)

curveIn = GafferSceneTest.CompoundObjectSource()
curveIn["in"].setValue(
	IECore.CompoundObject( {
		"bound" : IECore.Box3fData( testObject.bound() ),
		"children" : {
			"curveGroup" : {
				"bound" : IECore.Box3fData( testObject.bound() ),
				"children" : {
					"curves" : {
						"bound" : IECore.Box3fData( testObject.bound() ),
						"object" : testObject,
					},
				},
			},
		},
	} ),
)

root.addChild( curveIn )
```

mattigruener commented:
> It might be that a sensible solution is to give the option of either entering the text into the PythonEditor, or creating a variable to refer to the data instead?

Let's make that a separate issue/ticket?

johnhaddon commented:
Thanks Matti! 



> > It might be that a sensible solution is to give the option of either entering the text into the PythonEditor, or creating a variable to refer to the data instead?

> Let's make that a separate issue/ticket?

For sure, or we just keep our mouths shut and see if it comes up in practice :)
